### ISSUE TITLE
Performance regression in module support

### ISSUE THREAD

**Issue by bwester** [securego/gosec#406]
### Summary 

There appears to be a performance regression related to scanning packages. I've noticed
a drastic slowdown in runtime starting with commit 85d180848d "Go modules support for 1.12"
when run on some of the packages that I work with. Scan times were improved considerably
after commit a65402bc5a, which updated golang.org/x/tools and picked up 7dd31af5a6
"go/packages: allow types loading without NeedDeps", but it's still 3x slows that before
module support.

I've looked into the code myself, and I think the slowdown is due to how arguments are
scanned and loaded. Currently, each argument is scanned with `filepath.Walk` to get a
list of packages, then each package is passed through `build.Context.ImportDir` and
`packages.Load`. The `Load` function in turn just calls `go list`, and this call just
looks slow.

I suspect that repeatedly loading packages with lots of dependencies throws away repeated
work between executions. One way to speed this up might be to call `packages.Load` with
the all command line arguments at once, without the custom directory scanning and build
loading. I'd be happy to make a PR myself to do this if this sounds like a good approach.

### Steps to reproduce the behavior

```
time gosec ./...
```

I haven't been able to characterize exactly what makes a package slower. I suspect it has
to do with loading a large number of package, each having a large number of related
dependencies. I found this regression looking at timeouts in a CI pipeline on internal
packages. For a public example, `github.com/square/p2` exhibits this slowdown.

### gosec version

1.30 => ok
eaba99d => baseline: ok
85d1808 => horrible
2.0.0 => horrible
v2.1.0 => slow

### Go version (output of 'go version')

```
$ go version
go version go1.12.7 darwin/amd64
$ go env
GOARCH="amd64"
GOBIN=""
GOCACHE="..."  #redacted
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="..."  #redacted
GOPROXY=""
GORACE=""
GOROOT="/usr/local/opt/go/libexec"
GOTMPDIR=""
GOTOOLDIR="/usr/local/opt/go/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=... -gno-record-gcc-switches -fno-common"  #redacted
```

### Operating system / Environment

macOS Mojave

### Expected behavior

Experimental process: ran `time gosec ./...` from within `github.com/square/p2/bin`.
Reporting the average of 5 executions.

Environment: go1.12.7 with GO111MODULE=off

Git revision | Runtime
------------ | -------
eaba99d (baseline) | 4.04s

### Actual behavior

Environment: go1.12.7 with GO111MODULE=off

Git revision   | Runtime  | Relative Slowdown
-------------- | -------- | -----------------
commit 85d1808 | 41.01s   | 10.7 x
2.0.0          | 41.75    | 10.3 x
v2.1.0         | 12.09 (*)   |  3.0 x

*: The first execution still takes ~40s like the other commits, but there now looks to be a cache that can be reused for subsequent executions.

With an updated environment: go1.13.3 with GO111MODULE=on

Git revision   | Runtime  | Relative Slowdown
-------------- | -------- | -----------------
v2.1.0         | 11.79s   |  2.9 x

**Comment by ccojocar** [comment]
Yeah, ` GO111MODULE=off` is pretty slow. Most of the time the dependencies are required to have proper syntax checking. We used to just load individual files without dependencies but we had a number of issues due to not loading the dependencies or the entire package.

You can go ahead and propose an improvement in a pull request, but notice that `gosec` supports multiple paths and also some folders can be excluded from scanning.

**Comment by bwester** [comment]
Throwing all CLI arguments through `packages.Load` is straightforward and generally works, but I acknowledge that it produces a `gosec` command that will not have the same behavior as the current version. I haven't yet hit on the right logic allowing this new loading call while also loading the same kinds of arguments that the current version does. There is a difference between handling _files_/_directories_ and handling _packages_. The tricky cases I found are:

1. Excluding packages _before_ loading them
2. Handling packages with errors
3. Handling individual source files from different packages
4. Analyzing independent source files that aren't importable, e.g., because they aren't in a module or GOPATH

I've been hacking at this on and off over the last week, and I've spent all the time I've budgeted for this endeavor without success. I'm comfortable declaring my initial suggestion a dead end due to not being backwards compatible.
